src/components/ApartmentHallSelection/index.js

```diff
@@ -0,0 +1,132 @@
+    if (index !== null) {
+      // Send the selected index to the parent component
+      this.props.onHallRemove(index);
+    }
+  };
+
+  handleAddDropdown = () => {
+    this.props.onHallAdd();
+  };
+
+  handleSaveButtonClick = () => {
+    this.props.onSaveButtonClick();
+  };
+
+  render() {
+    return (
+      <Card>
+        <CardHeader title="Preferred Halls" className="card-header" />
+        <CardContent>
+          <Grid container justify="space-between" spacing={2}>
+            <Grid item xs={11}></Grid>
+            <Grid item xs={12}>
+              <List className="hall-list" aria-label="apartment preferred halls">
+                {this.props.preferredHalls ? (
+                  this.props.preferredHalls.map((preferredHall, index) => (
+                    <HallListItem
+                      key={preferredHall.hallName + index}
+                      index={index}
+                      availableHalls={this.state.availableHalls}
+                      primaryUsername={this.props.primaryUsername}
+                      preferredHalls={this.props.preferredHalls}
+                      onHallInputChange={this.handleInputChange}
+                      onHallRemove={this.handleRemove}
+                      authentication={this.props.authentication}
+                    />
+                  ))
+                ) : (
+                  <HallListItem
+                    key={''}
+                    index={0}
+                    availableHalls={this.state.availableHalls}
+                    primaryUsername={this.props.primaryUsername}
+                    preferredHalls={this.props.preferredHalls}
+                    onHallInputChange={this.handleInputChange}
+                    authentication={this.props.authentication}
+                  />
+                )}
+              </List>
+            </Grid>
+            <Grid item xs={12}>
+              <Button
+                variant="contained"
+                color="default"
+                startIcon={<AddIcon />}
+                onClick={this.handleAddDropdown}
+              >
+                Add a Hall
+              </Button>
+            </Grid>
+            <Grid item xs={9}>
+              {this.props.saving === 'failed' ? (
+                <Typography variant="overline" color="error">
+                  Something when wrong while trying to save the application
+                </Typography>
+              ) : this.props.preferredHalls.length >= this.state.availableHalls.length ? (
+                <Typography variant="overline" color="error">
+                  You have reached the maximum number of halls ({this.state.availableHalls.length})
+                </Typography>
+              ) : null}
+            </Grid>
+            <Grid item xs={3}>
+              <SaveButton saving={this.props.saving} onClick={this.handleSaveButtonClick} />
+            </Grid>
+          </Grid>
+        </CardContent>
+      </Card>
+    );
+  }
+}
```

Comment:
Here's a more substantial example of refactoring into a functional component, where the useState and useEffect hooks are needed.
```suggestion
import React, { useState, useEffect } from 'react';
import { Grid, Card, CardHeader, CardContent, List, Typography, Button } from '@material-ui/core';
import AddIcon from '@material-ui/icons/Add';
import HallListItem from './components/HallListItem';
import SaveButton from '../ApartAppSaveButton';
import goStalk from '../../services/goStalk';
import housing from '../../services/housing';
import '../../views/ApartmentApp/apartmentApp.css';
import '../../views/PeopleSearch/components/PeopleSearchResult/peopleSearchResult.css';

// Create a list of selection boxes to choosing preferred halls
const HallSelection = ({
  authentication,
  saving,
  primaryUsername,
  preferredHalls,
  onHallInputChange,
  onHallRemove,
  onHallAdd,
  onSaveButtonClick,
}) => {
  const [halls, setHalls] = useState([]);

  useEffect(() => {
    const loadHalls = async () => {
      let unfilteredHalls;
      try {
        // Get the halls available for apartments, filtered by the gender of the primary applicant
        unfilteredHalls = await housing.getApartmentHalls(primaryUsername);
      } catch {
        //! DEBUG: Fills in halls dropdown when the housing api endpoint is not yet implemented
        // This list of halls is references from the 'Hall' dropdown on the PeopleSearch page
        unfilteredHalls = await goStalk.getHalls();
      }
      //Remove spaces from strings
      let availableHalls = unfilteredHalls.map((hall) => hall.trim());
      setHalls(availableHalls);
    };

    if (authentication) {
      loadHalls();
    }
  }, [authentication, primaryUsername]);

  const handleInputChange = (hallSelectionValue, hallRankValue, index) => {
    console.log('Called "handleInputChange" in ApartmentHallSelection component');
    console.log('hallName: ' + hallSelectionValue); //! DEBUG
    console.log('hallRank: ' + hallRankValue); //! DEBUG
    console.log('index: ' + index); //! DEBUG
    onHallInputChange(hallSelectionValue, hallRankValue, index);
  };

  const handleRemove = (index) => {
    // Make sure the chosen index was not null
    if (index !== null) {
      // Send the selected index to the parent component
      onHallRemove(index);
    }
  };

  return (
    <Card>
      <CardHeader title="Preferred Halls" className="card-header" />
      <CardContent>
        <Grid container justify="space-between" spacing={2}>
          <Grid item xs={11}></Grid>
          <Grid item xs={12}>
            <List className="hall-list" aria-label="apartment preferred halls">
              {preferredHalls ? (
                preferredHalls.map((preferredHall, index) => (
                  <HallListItem
                    key={preferredHall.hallName + index}
                    index={index}
                    availableHalls={halls}
                    primaryUsername={primaryUsername}
                    preferredHalls={preferredHalls}
                    onHallInputChange={handleInputChange}
                    onHallRemove={handleRemove}
                    authentication={authentication}
                  />
                ))
              ) : (
                <HallListItem
                  key={''}
                  index={0}
                  availableHalls={halls}
                  primaryUsername={primaryUsername}
                  preferredHalls={preferredHalls}
                  onHallInputChange={handleInputChange}
                  authentication={authentication}
                />
              )}
            </List>
          </Grid>
          <Grid item xs={12}>
            <Button variant="contained" color="default" startIcon={<AddIcon />} onClick={onHallAdd}>
              Add a Hall
            </Button>
          </Grid>
          <Grid item xs={9}>
            {saving === 'failed' ? (
              <Typography variant="overline" color="error">
                Something when wrong while trying to save the application
              </Typography>
            ) : preferredHalls.length >= halls.length ? (
              <Typography variant="overline" color="error">
                You have reached the maximum number of halls ({halls.length})
              </Typography>
            ) : null}
          </Grid>
          <Grid item xs={3}>
            <SaveButton saving={saving} onClick={onSaveButtonClick} />
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  );
};

export default HallSelection;
```

Reply:
Thank you, this is very helpful